### ISSUE TITLE
Possibility to give Widgets a name in the view hierarchy (so the Play Store "Pre-launch report" Sign-in credentials can be configured)

### ISSUE THREAD

**Issue by hugovangalen** [flutter/flutter#16921]
I was looking into the "Pre-launch Report" Sign in credentials configuration in the Play Store console. 

I pulled open Android Device Monitor while running the app on the login-screen, and did not expect to get a valid view XML snapshot returned, that's actually pretty nice work! 

It appears however that it is not possible to give a widget a 'resource name', as required by the pre-launch report Sign-in credentials configuration. 

Will it become possible at some point to set this value? I guess once its there, it would allow developers to use all kinds of automated testing tools on their apps.

**Comment by Hixie** [comment]
Not sure exactly what we have to do to support this.

https://support.google.com/googleplay/android-developer/answer/7002270?hl=en

cc @eseidelGoogle @mit-mit

**Comment by lyingju** [comment]
Any update on this? It would be great if flutter app can leverage the tools in Google play console

**Comment by eseidelGoogle** [comment]
@mit-mit do you still manage our publishing of Flutter Gallery to Play?  Can you comment on what we are missing here?

**Comment by matthewcook333** [comment]
![image](https://user-images.githubusercontent.com/4873536/49678635-90409980-fa3a-11e8-9a7e-2ec7058a9985.png)

This is what the screen to fill in looks like, and the ? by "Username resource name" says "The Android resource name of the text field within your app where the given username should be entered."

**Comment by eseidelGoogle** [comment]
Ah!  This probably falls into then discussions we've had around making sure Flutter plays nice with roboelectric, etc.  FYI @goderbauer @gspencergoog @mehmetf

**Comment by jonahwilliams** [comment]
The issue with annotating widgets with ids is that a single accessibility node may be composed of many flutter widgets.

**Comment by mehmetf** [comment]
There's a semi-answer from Android team and it does not look good. It seems much of that tooling uses native view hierarchy so it is incompatible with Flutter. If this becomes a priority, we need to work with them to support interacting with accessibility layer.

**Comment by jsalhi** [comment]
This also would help with automation testing in [Firebase Test Lab](https://firebase.google.com/docs/test-lab/), since their supported testing frameworks ([Robotium](https://github.com/RobotiumTech/robotium), [UI Automator](https://developer.android.com/training/testing/ui-automator), [Espresso](https://developer.android.com/training/testing/espresso/)) rely on the native view hierarchy and being able to specify elements by their resource id. Currently it does not seem possible to use their integration testing platform with Flutter apps.

**Comment by neiljaywarner** [comment]
could tis be helpful - findbytag?
https://github.com/appium/appium/issues/6025

**Comment by manuelgomes2** [comment]
What about the Semantic() class, can this be used to add/find ID's?
Google's pre-launch report is already able to find the inputs - in the image below (a Flutter app's login screen), it is able to find the username and password fields:
![image](https://user-images.githubusercontent.com/7493585/54058893-27122300-41ef-11e9-8df0-df8e6d4db8c8.png)

Could this be a way to go? 
This does not seem to be impossible

**Comment by jld3103** [comment]
It would be really nice to have such a feature.

**Comment by fvisticot** [comment]
We need this feature ! Testing an app is not only to test the login screen ...

**Comment by xiyulangzi** [comment]
yes, we need this feature. Google Play console team and Flutter team need work tegother.

**Comment by eseidelGoogle** [comment]
TIL that appbundles may also break the pre-launch report? https://github.com/flutter/flutter/issues/18494#issuecomment-503669889

**Comment by eseidelGoogle** [comment]
FYI @collinjackson

**Comment by sceee** [comment]
I am actually having the issue that Google Play Pre-Launch Report cannot go past my Flutter app's initial wizard that is implemented as a Pager. It just never scrolls to the next page. Is this a separate issue or the same?

**Comment by jld3103** [comment]
@sceee It's a different issue. This issue is about naming widgets in the view hierarchy for supplying the credentials in the PlayStore.

**Comment by sceee** [comment]
Thanks @jld3103 , I opened a separate issue for it.

**Comment by xiyulangzi** [comment]
@eseidelGoogle any update for this case?

**Comment by eseidelGoogle** [comment]
I have no updates, sorry.

Sounds like @mehmetf got some pointers from the Android team, but we also likely need to reach out to the play team to learn more about how this is implemented and if it's something which can be made work with Flutter.  FYI @csells

**Comment by smrucv** [comment]
Hi,

any update about this?

Plz.

**Comment by Dararii** [comment]
any update on this?

**Comment by deandreamatias** [comment]
Please stop commenting on problems

> any update on this
The best option is to add this emoji 👍 in the initial comment.

To see a priority of each problem, we can see in which Milestone the isse is added. This issue belongs 
to milestone Stretch Goals, with this description:

>  Due by January 01, 2032 6% complete
> P3: This milestone is for features and bugs that we feel we could easily defer until after those in the "Goals" milestone. We are unlikely to work on these issues in the near future. The date on this milestone is utterly arbitrary and just intended to sort the milestone appropriately.

**Comment by Ibtesam-Mahmood** [comment]
Any updates on this?

**Comment by eseidelGoogle** [comment]
No updates.  My understanding is that Pre-launch Report (PLR) is many tools all working together under an umbrella framework.  Some of those tools are designed around assumptions of working with Android Views/Espresso.  We should probably be clear which of the PLR tools do/don't work well with Flutter.  I've also put @blasten in touch with a member of the Pre-Launch Report team in case there are some easy wins here.

**Comment by Project-AssetDiary** [comment]
Pre Launch Tool doesn't work well with Flutter. I hope to support this feature as soon as possible.

**Comment by PhilipPurwoko** [comment]
I done it using widget key for Google Play. Just like below

```dart
CustomTextFormField(
  key: const Key('username'),
  controller: _usernameController,
  labelText: 'Username',
  textInputAction: TextInputAction.next,
  inputFormatters: [
    FilteringTextInputFormatter.deny(RegExp(r'\s+\b|\b\s|\s|\b')),
  ],
  validator: (String? value) {
    if (value == null || value.isEmpty) {
      return 'Username cannot be empty';
    }
    return null;
  },
),
verticalSpaceRegular,
CustomTextFormField(
  key: const Key('password'),
  controller: _passwordController,
  obscureText: _obscureText,
  labelText: 'Password',
  useCustomSuffix: true,
  textInputAction: TextInputAction.send,
  suffixIcon: IconButton(
    icon: Icon(_obscureText
        ? Icons.remove_red_eye
        : Icons.remove_red_eye_outlined),
    onPressed: () => setState(() {
      _obscureText = !_obscureText;
    }),
  ),
  validator: (String? value) {
    if (value == null || value.isEmpty) {
      return 'Password cannot be empty';
    }
    return null;
  },
),
```

**Comment by eltonmorais** [comment]
> 

@PhilipPurwoko and them you filled the resource name field only with the "key"? Or used anything else? Have you tested it again?

**Comment by lisovyk** [comment]
@PhilipPurwoko would appreciate a MRE if you'd be so kind; seems like many people have this issue, myself included.

**Comment by PhilipPurwoko** [comment]
> I done it using widget key for Google Play. Just like below
> 
> ```dart
> CustomTextFormField(
>   key: const Key('username'),
>   controller: _usernameController,
>   labelText: 'Username',
>   textInputAction: TextInputAction.next,
>   inputFormatters: [
>     FilteringTextInputFormatter.deny(RegExp(r'\s+\b|\b\s|\s|\b')),
>   ],
>   validator: (String? value) {
>     if (value == null || value.isEmpty) {
>       return 'Username cannot be empty';
>     }
>     return null;
>   },
> ),
> verticalSpaceRegular,
> CustomTextFormField(
>   key: const Key('password'),
>   controller: _passwordController,
>   obscureText: _obscureText,
>   labelText: 'Password',
>   useCustomSuffix: true,
>   textInputAction: TextInputAction.send,
>   suffixIcon: IconButton(
>     icon: Icon(_obscureText
>         ? Icons.remove_red_eye
>         : Icons.remove_red_eye_outlined),
>     onPressed: () => setState(() {
>       _obscureText = !_obscureText;
>     }),
>   ),
>   validator: (String? value) {
>     if (value == null || value.isEmpty) {
>       return 'Password cannot be empty';
>     }
>     return null;
>   },
> ),
> ```

False. I have tried it again and it is not working. I use the other solution that I have - commented above

**Comment by davymont** [comment]
@PhilipPurwoko where do you find the app content > app access etc. that you're talking about?

**Comment by PhilipPurwoko** [comment]
@davymont Menu in Google Play Console

**Comment by RicardoRB** [comment]
Something new here?

**Comment by maks** [comment]
It seems like our Compose friends [are in the same boat](https://issuetracker.google.com/issues/220801884?pli=1) so perhaps if they come up with a solution it could be used as a basis for a solution for Flutter as well 🤞🏻.

**Comment by MahendraRaj94** [comment]
One can explore more about `Semantics` widget in Flutter to see if it provides a solution. 

There could be a possible solution to assign *resource identifiers* to widgets in **`Flutter`** using **`Semantics`** widget. 

`Semantics` widget has an attribute called **`identifier`** which we can set as per the [Flutter documentation](https://api.flutter.dev/flutter/semantics/SemanticsProperties/identifier.html).

Again, I haven't tested it but it seems it helps in UI testing.
